File: src/main/java/com/gdep/PackScore.java
package com.gdep;

import com.gdep.FuzzyMatch.FuzzyMatchResult;
import java.util.List;

public class PackScore {
    public static int scoreClassNameSimilarity(String filePath, String className) {
        if (filePath.length() <= 0 || className.length() <= 0) {
            return Integer.MIN_VALUE;
        }

        // normalize slash
        filePath = filePath.replace("\\", "/");

        // remove file extension from filePath
        {
            int slashIndex = filePath.lastIndexOf('/');
            if (slashIndex < 0) {
                slashIndex = 0;
            }

            int dotIndex = filePath.indexOf('.', slashIndex + 1);
            if (dotIndex >= 0) {
                filePath = filePath.substring(0, dotIndex);
            }
        }

        List<String> filePathParts = Util.plainSplit(filePath, "/");
        List<String> classNameParts = Util.plainSplit(className, ".");

        int score = 0;

        String filePathAbbreviated = getAbbreviated(filePathParts);
        String classNameAbbreviated = getAbbreviated(classNameParts);

        score -=
                FuzzyMatch.fuzzyMatch(filePathAbbreviated, classNameAbbreviated).distance() * 10;

        String filePath2 = filePath.replace("/", ".");
        String className2 = removeLogNoise(classNameParts);

        {
            String str = filePath2.length() > className2.length() ? filePath2 : className2;
            String sub = filePath2.length() > className2.length() ? className2 : filePath2;

            FuzzyMatchResult res = FuzzyMatch.fuzzyMatch(str, sub);

            score -= res.distance() * 10;
            score -= Math.abs(str.length() - res.length());
        }

        return score;
    }

    private static String getAbbreviated(List<String> parts) {
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < parts.size(); i++) {
            String part = parts.get(i);

            if (!part.isEmpty()) {
                sb.append(Character.toLowerCase(part.charAt(0)));
            }

            if (i + 1 < parts.size()) {
                sb.append(".");
            }
        }

        return sb.toString();
    }

    private static String removeLogNoise(List<String> classNameParts) {
        StringBuilder sb = new StringBuilder();

        boolean abbrevEnded = false;

        for (int i = 0; i < classNameParts.size(); i++) {
            String part = classNameParts.get(i);

            if (part.length() > 1) {
                abbrevEnded = true;
            }

            if (!abbrevEnded) {
                continue;
            }

            if (i == classNameParts.size() - 1) { // last part
                int dollarIndex = part.indexOf("$");

                if (dollarIndex >= 0) {
                    part = part.substring(0, dollarIndex);
                }
                sb.append(part);
            } else {
                sb.append(part);
                sb.append(".");
            }
        }

        return sb.toString();
    }
}

File: go_wrapper/pack_score.go
package main

import (
	"math"
	"strings"
	"unicode"
	"unicode/utf8"
)

func ScoreClassNameSimilarity(
	filePath, className string,
	mat *Matrix,
) int {
	if filePath == "" || className == "" {
		return math.MinInt
	}

	// normalize slash
	filePath = strings.ReplaceAll(filePath, "\\", "/")

	// remove file extension from filePath
	{
		slashIndex := strings.LastIndexByte(filePath, '/')
		if slashIndex < 0 {
			slashIndex = 0
		}

		dotIndex := -1
		for i := slashIndex + 1; i < len(filePath); i++ {
			if filePath[i] == '.' {
				dotIndex = i
				break
			}
		}

		if dotIndex >= 0 {
			filePath = filePath[0:dotIndex]
		}
	}

	filePathParts := strings.Split(filePath, "/")
	classNameParts := strings.Split(className, ".")

	score := 0

	filePathAbbreviated := GetAbbreviated(filePathParts)
	classNameAbbreviated := GetAbbreviated(classNameParts)

	score -= FuzzyMatch(filePathAbbreviated, classNameAbbreviated, mat).Distance * 10

	filePath2 := strings.ReplaceAll(filePath, "/", ".")
	className2 := RemoveLogNoise(classNameParts)

	{
		var str string
		var sub string

		if len(filePath2) > len(className2) {
			str = filePath2
			sub = className2
		} else {
			str = className2
			sub = filePath2
		}

		res := FuzzyMatch(str, sub, mat)

		score -= res.Distance * 10
		score -= int(AbsInt(len(str) - res.Length()))
	}

	return score
}

func GetAbbreviated(parts []string) string {
	sb := strings.Builder{}

	for i := 0; i < len(parts); i++ {
		part := parts[i]

		r, _ := utf8.DecodeRuneInString(part)
		if r != utf8.RuneError {
			sb.WriteRune(unicode.ToLower(r))
		}

		if i+1 < len(parts) {
			sb.WriteString(".")
		}
	}

	return sb.String()
}

func RemoveLogNoise(classNameParts []string) string {
	sb := strings.Builder{}

	abbrevEnded := false

	for i := 0; i < len(classNameParts); i++ {
		part := classNameParts[i]

		if len(part) > 1 {
			abbrevEnded = true
		}

		if !abbrevEnded {
			continue
		}

		if i == len(classNameParts)-1 {
			dollarIndex := strings.IndexByte(part, '$')

			if dollarIndex >= 0 {
				part = part[0:dollarIndex]
			}
			sb.WriteString(part)
		} else {
			sb.WriteString(part)
			sb.WriteString(".")
		}
	}

	return sb.String()
}
